D2k inaccuracy

penev92 opened the pull request:
This implements the first point of the TODO list in #17972 - it adds distinct inaccuracy calculation types to projectiles and switches weapons in the D2k mod to the new calculation type (while also adjusting the inaccuracy values to account for it).

P.S.: I didn't know where to put the InaccuracyType  enum so I just plopped it anywhere.

penev92 committed:
Added syncing to some Railgun projectile fields

penev92 committed:
Added InaccuracyType enum and updated projectiles accordingly

Also updated the inaccuracy calculations to account for the new inaccuracy type - either based on distance up to a max defined inaccuracy at max range (old style) or based on distance with each cell (1024 range) increasing the inaccuracy with a set step.

penev92 committed:
Changed weapons in D2k to use the new PerCellIncrement inaccuracy

Also adjusted the inaccuracy values. This should bring inaccuracy in D2k pretty much in line with the original game, with the potential liberty of decreased inaccuracy for the Deviator tank.

abcdefg30 commented on mods/d2k/weapons/largeguns.yaml:
```diff
@@ -89,7 +89,7 @@ DevBullet:
 		Shadow: true
 		LaunchAngle: 62
-		Inaccuracy: 768
+		Inaccuracy: 256 # The siege tank of the original somehow doesn't seem to be any more inaccurate than a normal combat tank, but following the ORA tradition we increase the inaccuracy because this is an artillery unit.
```
Can we remove this comment, please?

penev92 replied:
We can, but if I remove it I want to first have a final decision that we're going to have increased inaccuracy on the Siege tank compared to the original.

pchote replied:
we should leave it to a future balance pass, IMO.

penev92 replied:
OK, will set it to match the original then.

reaperrr replied:
Yeah, I'd stick with the original value for now (same then goes for the Spread, in case you increased that to make up for the higher inaccuracy).

penev92 replied:
Removed the override along with the comment.

penev92 commented:
I will look into changing the inaccuracy logic a bit, so in the mean time reviews are probably best done elsewhere.

penev92 committed:
Added InaccuracyType.Absolute to projectiles

penev92 commented:
Added another inaccuracy type - `Absolute` and switched the default for `MissileProjectile` to that to retain the current behavior.
Also removed inaccuracy overrides for the Siege tank and Missile tank. Only inaccuracy override I took liberties with was the Deviator tank to make sure it actually affects its target.

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -121,15 +125,29 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
-				var maxOffset = inaccuracy * (target - headPos).Length / args.Weapon.Range.Length;
+
+				var maxOffset = 0;
+				switch (info.InaccuracyType)
```
Please pull this out into a helper method instead of duplicating it everywhere!

penev92 replied:
This was my original intention as well, but adding a base class and all the nice things that come with it makes this a lot harder to review, so I opted for doing it in a follow-up PR.

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -21,6 +21,8 @@
 
 namespace OpenRA.Mods.Common.Projectiles
 {
+	public enum InaccuracyType { Maximum, PerCellIncrement, Absolute }
```
This should then be moved to wherever the helper method goes (probably `OpenRA.Mods.Common/Util.cs`

pchote commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -47,9 +49,11 @@ public class AreaBeamInfo : IProjectileInfo
 		[Desc("Ranges at which each Falloff step is defined.")]
 		public readonly WDist[] Range = { WDist.Zero, new WDist(int.MaxValue) };
 
-		[Desc("Maximum offset at the maximum range.")]
```
Can we change this to refer to `InaccuracyType` instead of removing it? Likewise in the other projectiles.

pchote commented on OpenRA.Mods.Common/Projectiles/Railgun.cs:
```diff
@@ -108,6 +110,11 @@ public class Railgun : IProjectile
 		bool animationComplete;
+
+		[Sync]
+		WPos source;
```
We don't appear to be using this anywhere?

penev92 replied:
You're right - it's only used in the constructor.

penev92 committed:
Moved projectile inaccuracy calculations to Common.Util

Also moved the InaccuracyType enum there. This also quietly adds the RangeModifiers to the calculations for all projectiles, while they were only used on Bullet so far, which seemed very wrong.

penev92 committed:
Added back the DescAttributes on projectiles' Inaccuracy fields

penev92 commented:
Rebased for good measure and addressed the above comments.
Note that this PR "secretly" fixes a few side things:
 - Adds `SyncAttribute` to the `Railgun` projectile's `target` field, which seems like it was mistakenly missed.
 - Adds inaccuracy calculations to the `Railgun` projectile, which had an `Inaccuracy` property but wasn't using it.
 - Adds an option for the `Missile` projectile to use the same inaccuracy method all other projectiles used (so far `Missile`'s inaccuracy would be constant regardless of range, hence the `Absolute` inaccuracy type). Now all projectiles can use all inaccuracy methods.
 - Adds the `RangeModifiers` to the calculations for all projectiles, while they were only used on `Bullet` so far, which seemed very wrong.

reaperrr commented on OpenRA.Mods.Common/Projectiles/AreaBeam.cs:
```diff
@@ -125,7 +125,7 @@ public AreaBeam(AreaBeamInfo info, ProjectileArgs args, Color color)
 				speed = info.Speed[0];
 
-			// Both the head and tail start at the source actor, but initially only the head is travelling.
+			// Both the head and tail start at the source actor, but initially only the head is traveling.
```
Minor nit: two `l` is correct for British English and less common, but also not wrong for American English.
No blocker, but in case this PR needs another update, imo just drop this while you're at it.

penev92 replied:
Sorry, my spellchecker was complaining and I didn't think about that. Reverted.

reaperrr reviewed (dismissed):
Looks good to me 👍

pchote commented:
A couple of the commits here have a subject line that overrun the length limit. Can you please reword those to be shorter?

penev92 commented:
Reworded commit messages and rebased for good measure.

abcdefg30 commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -228,5 +230,22 @@ public static string FriendlyTypeName(Type t)
+				case InaccuracyType.Absolute:
+					return inaccuracy;
+				default:
+					return 0;
```
This case should never be hit, right? I assume this is then to silence some sort of compiler warning. Wouldn't it make more sense to throw an exception instead?

pchote replied:
IMO it would be reasonable to default unknown `InaccuracyType`s to `InaccuracyType.Absolute`, so we could just add the `default` label to that case?

penev92 replied:
Having the default of the TraitInfo field be `Maximum` but then defaulting this to `Absolute` would be weird.

pchote replied:
@abcdefg30 can you confirm that this resolves your question or suggest a suitable fix so we can move this forward?

abcdefg30 replied:
I thought the result here was defaulting to `Maximum` since that is the default on the info field?

penev92 replied:
As you initially assumed, this is only here to please the compiler, because otherwise we get an error (not a warning) - `not all code paths return a value`. So one way or another we need a `default` case or a `return` after the `switch` (or to throw an exception).
`InaccuracyType.Absolute` might be the most logical default, but as I pointed out, it needs to be the default in both places and one of them would require a major update rule that really isn't needed.
The `default` here results in no inaccuracy, actually. So if one sets up their `InaccuracyType` incorrectly they will get no inaccuracy ingame. Hopefully that would be enough to make them realize the error without having to throw an exception.

pchote replied:
Can we compromise with a `throw new InvalidEnumArgumentException("inaccuracyType", (int)inaccuracyType, typeof(InaccuracyType));` so we can merge this and move on to more productive arguments?

penev92 replied:
I have nothing against throwing an exception if people are fine with that.

penev92 commented:
Updated with the `throw`.

teinarss commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -11,16 +11,18 @@
 
 namespace OpenRA.Mods.Common
 {
+	public enum InaccuracyType { Maximum, PerCellIncrement, Absolute }
```
Does this really belong to the util file? :D

penev92 replied:
Same answer as below. :(

teinarss commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -228,5 +230,22 @@ public static string FriendlyTypeName(Type t)
 			return t.Name;
 		}
+
+		public static int GetProjectileInaccuracy(int baseInaccuracy, InaccuracyType inaccuracyType, ProjectileArgs args)
```
Also im not a really big fan of using a Utility class where we can just throw stuff into that we have hard to find a home for.

penev92 replied:
That is exactly what I said, but since we have no better place I was told by @pchote to put it here. I was considering making a base projectile class, but that was out of scope.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9083edc67c6ade0b4e9fba3338e0ae45365447a8...7ea10fef4cc4b970b62472fe96d4e2dbc2fac383)
